Add simple in-memory Window function implementation

aditi-pandit opened the pull request.

aditi-pandit commented:
Changes in progress : Basic window frame functionality.

mbasmanova commented on velox/exec/Window.cpp:
```diff
@@ -0,0 +1,465 @@
+}
+
+RowVectorPtr Window::getOutput() {
+  if (finished_ || !noMoreInput_ || returningRows_.size() == numRowsApplied_) {
```
Can we remove 'returningRows_.size() == numRowsApplied_' check by adding logic to set finished to true when we are done processing?

aditi-pandit replied:
That existed already. Removing this condition.

mbasmanova commented on velox/exec/Window.cpp:
```diff
@@ -0,0 +1,465 @@
+        result->childAt(i));
+  }
+
+  std::vector<VectorPtr> windowFunctionOutputs;
```
No need for this vector. Use result instead.

aditi-pandit replied:
I have used this vector in intermediate computations.

mbasmanova commented on velox/functions/prestosql/window/RowNumber.cpp:
```diff
@@ -0,0 +1,78 @@
+      const BufferPtr& /*frameStarts*/,
+      const BufferPtr& /*frameEnds*/,
+      int32_t /*startPartitionRow*/,
+      int32_t resultIndex,
```
perhaps, rename this to resultOffset or offset

mbasmanova commented on velox/functions/prestosql/window/RowNumber.cpp:
```diff
@@ -0,0 +1,78 @@
+          const std::vector<TypePtr>& argTypes,
+          const TypePtr& /*resultType*/)
+          -> std::unique_ptr<exec::WindowFunction> {
+        VELOX_CHECK_EQ(argColumns.size(), 0, "{} takes no arguments", name);
```
These checks are redundant as the framework already knows the signature and can do these checks in a generic manner. Let's remove them.

mbasmanova commented on velox/functions/prestosql/window/RowNumber.cpp:
```diff
@@ -0,0 +1,78 @@
+  return true;
+}
+
+static bool FB_ANONYMOUS_VARIABLE(g_WindowFunction) =
```
Aggregate functions are registered this way, but it is quite problematic. Let's follow the mechanism used for scalar functions which requires explicit register call.

mbasmanova commented on velox/functions/prestosql/window/Rank.cpp:
```diff
@@ -0,0 +1,86 @@
+    auto* rawResultsVector =
+        result->asFlatVector<int64_t>()->mutableRawValues();
+    for (int i = 0; i < numRows; i++) {
+      if (peerGroupStarts->as<size_t>()[i] != currentPeerGroupStart_) {
```
use temp variable for peerGroupStarts->as<size_t>()

mbasmanova commented on velox/functions/prestosql/window/NthValue.cpp:
```diff
@@ -0,0 +1,127 @@
+
+  void resetPartition(const std::vector<char*>& rows) {
+    partitionRows_ = reinterpret_cast<char* const*>(rows.data());
+    /*firstArgVector_->resize(rows.size());
```
remove commented out code

mbasmanova commented on velox/functions/prestosql/window/NthValue.cpp:
```diff
@@ -0,0 +1,127 @@
+        partitionRows_(nullptr) {}
+
+  void resetPartition(const std::vector<char*>& rows) {
+    partitionRows_ = reinterpret_cast<char* const*>(rows.data());
```
Why do you need to reinterpret_cast?

mbasmanova commented on velox/functions/prestosql/window/NthValue.cpp:
```diff
@@ -0,0 +1,127 @@
+      int32_t startPartitionRow,
+      int32_t resultIndex,
+      const VectorPtr& result) {
+    FlatVector<T>* resultVector = result->as<FlatVector<T>>();
```
auto

mbasmanova commented on velox/functions/prestosql/window/NthValue.cpp:
```diff
@@ -0,0 +1,127 @@
+      int32_t resultIndex,
+      const VectorPtr& result) {
+    FlatVector<T>* resultVector = result->as<FlatVector<T>>();
+    auto firstArgVector =
```
re-use the vector across batches

mbasmanova commented on velox/functions/prestosql/window/NthValue.cpp:
```diff
@@ -0,0 +1,127 @@
+
+    int numRows = peerGroupStarts->size();
+    for (int i = 0; i < numRows; i++) {
+      exec::RowContainer::extractColumn(
```
Why do we extract values one row at a time? Can we extract all rows at once before the loop?

mbasmanova commented on velox/functions/prestosql/window/NthValue.cpp:
```diff
@@ -0,0 +1,127 @@
+          offsetsVector);
+      const int64_t offset = offsetsVector->values()->as<int64_t>()[0];
+      VELOX_CHECK_GE(offset, 1);
+      const int64_t frameStart = frameStarts->as<size_t>()[i];
```
use temp variable for frameStarts->as<size_t>()

int64_t -> vector_size_t

mbasmanova commented on velox/functions/prestosql/window/NthValue.cpp:
```diff
@@ -0,0 +1,127 @@
+      const int64_t offset = offsetsVector->values()->as<int64_t>()[0];
+      VELOX_CHECK_GE(offset, 1);
+      const int64_t frameStart = frameStarts->as<size_t>()[i];
+      VELOX_CHECK_GE(frameStart, 0);
```
remove this check from the hot loop

mbasmanova commented on velox/functions/prestosql/window/NthValue.cpp:
```diff
@@ -0,0 +1,127 @@
+      VELOX_CHECK_GE(offset, 1);
+      const int64_t frameStart = frameStarts->as<size_t>()[i];
+      VELOX_CHECK_GE(frameStart, 0);
+      exec::RowContainer::extractColumn(
```
Let's modify RowContainer API to allow extracting the value directly onto resultVector.

aditi-pandit replied:
I am no longer using this API.

mbasmanova commented on velox/functions/prestosql/window/NthValue.cpp:
```diff
@@ -0,0 +1,127 @@
+          1,
+          argColumns_[0],
+          firstArgVector);
+      resultVector->mutableRawValues()[resultIndex + i] =
```
This won't work for strings as we need to copy not only StringView, but the underlying string as well.

aditi-pandit commented on velox/exec/WindowFunction.cpp:
```diff
@@ -0,0 +1,69 @@
+} // namespace
+
+bool registerWindowFunction(
+    const std::string& name,
```
The name parameter is only used for lookup and not stored anywhere. So this needn't be done.

aditi-pandit commented on velox/exec/Window.cpp:
```diff
@@ -0,0 +1,465 @@
+    }
+  }
+  currentPartitionIndex_ = 0;
+  numberOfPartitions_ = j + 1;
```
I realized that this variable was unnecessary. Removed it.

aditi-pandit commented on velox/exec/WindowFunction.h:
```diff
@@ -0,0 +1,82 @@
+
+class WindowFunction {
+ public:
+  explicit WindowFunction(TypePtr resultType) : resultType_(resultType) {}
```
So far I haven't encountered a function that needed input columns that weren't arguments. However, we can change this API if needed subsequently.

aditi-pandit commented on velox/exec/Window.cpp:
```diff
@@ -0,0 +1,465 @@
+
+      auto frameEndPoints = findFrameEndPoints(
+          w, partitionStartRows_[currentPartitionIndex_], endRow, i);
+      VELOX_CHECK_EQ(frameEndPoints.second, i);
```
Removing this check. I'll also move all the checks in findFrameEndPoints to the constructor to fail early and reduce per-row logic.

aditi-pandit commented on velox/core/PlanNode.h:
```diff
@@ -57,6 +57,11 @@ class SortOrder {
   SortOrder(bool ascending, bool nullsFirst)
       : ascending_(ascending), nullsFirst_(nullsFirst) {}
 
+  SortOrder(const SortOrder& other) {
```
This was needed for the initKeyInfo function (in Window.cpp) when adding the default sort order into the vectors.

aditi-pandit commented on velox/exec/WindowFunction.h:
```diff
@@ -0,0 +1,82 @@
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      int32_t startPartitionRow,
+      int32_t resultIndex,
```
Yes, we can count the partition rows in the function implementation. I have removed that parameter.

aditi-pandit commented:
Squashed and cleaned the commits to https://github.com/facebookincubator/velox/pull/1816 for further reviews.

These review comments are also addressed in that PR. Lets continue further discussion there.
